docs/Codebase.md

```diff
@@ -0,0 +1,93 @@
+
+Besides the solution, the documentation and the `.yml` file for the build pipeline also reside in the repo at https://github.com/dotnet/insertions-client.
+
+This section covers the highlights of both `Core` and the `Console` projects as well as the `.yml` file for the build pipeline.
+
+**Table of Contents**
+- [Core Project](#core-project)
+  - [IInsertionApi Interface](#iinsertionapi-interface)
+  - [InsertionApi Class](#insertionapi-class)
+- [Console Project](#console-project)
+  - [Program Class](#program-class)
+  - [Logging](#logging)
+- [Build Pipeline](#build-pipeline)
+
+
+### [Core Project](#core-project)
+Core is the project where the main insertion logic resides in. The project exposes an interface (`IInsertionApi`) so that the code can be consumed by other projects and invoked like an API. 
+- `Console` project is an example of this. It receives the inputs from the console and invokes the api accordingly. See [Input](docs/README.md#Input) section for more details on how to use the tool from command line.
+
+#### [IInsertionApi Interface](#iinsertionapi-interface)
+Defines how to interact with the library to update versions of targeted assets, such as _default.config_.
+```csharp
+UpdateResults UpdateVersions(
+            IEnumerable<string> manifestFiles,
+            string defaultConfigFile,
+            IEnumerable<Regex> whitelistedPackages,
+            ImmutableHashSet<string>? packagesToIgnore,
+            string? accessToken,
+            string? propsFilesRootDirectory,
+            Predicate<Build>? buildFilter);
+```
+- __manifestFiles:__ The paths to all the manifest.json files to be inserted.
+- __defaultConfigFile:__ Full path to _default.config_ file.
+- __whitelistedPackages:__ Regex patterns matching with IDs of the packages that are allowed to be updated. If the set is empty,
+ any package is allowed be updated unless specified in packagesToIgnore argument.
+- __packagesToIgnore:__ IDs of the packages to be ignored.
+- __accessToken:__ Access token to be used when connecting to VS feed.
+- __propsFilesRootDirectory:__ Directory that will be searched for `.props` files.
+- __buildFilter:__ Predicate to determine if a build should be included in the insertion process.
+- Returns: an `UpdateResults` instance, detailing the outcome of the operation.
+
+In a simple case, this method can be invoked in the following way:
+```csharp
+ IInsertionApiFactory apiFactory = new InsertionApiFactory();
+ IInsertionApi api = apiFactory.Create(MaxWaitDuration, MaxDownloadDuration, MaxConcurrency);
+ api.UpdateVersions(new []{"\\Assets\\manifest.json"},
+    "\\\\VS\\.corext\\default.config",
+    Enumerable.Empty<Regex>(),
+    ImmutableHashSet.Create("VS.ExternalAPIs.MSBuild"),
+    null, null, null);
+```
+
+#### [InsertionApi Class](#insertionapi-class)
+`InsertionApi` is the only implementation of `IInsertionApi` in the solution. The main logic executes the following steps in order:
```

Comment:
Consider:  `InsertionApi` is the only class in `Core` that implements `IIInsertionApi`.

Reply:
Can you give a bit more info on the reasoning behind your comment? I can't tell in what way the proposed sentence is different.